File: Services/MousePerformanceCaptureService.vb
Imports System.Diagnostics
Imports System.Runtime.Versioning
Imports System.Threading
Imports WpfApp1.Models

Namespace Services
    <SupportedOSPlatform("windows")>
    Public Class MousePerformanceCaptureService
        Implements IDisposable

        Private Const DefaultPacketBufferCapacity As Integer = 16384

        Private ReadOnly _syncRoot As New Object()
        Private ReadOnly _queueProcessingSyncRoot As New Object()
        Private ReadOnly _rawInputBroker As IRawInputBroker
        Private ReadOnly _sessionRouter As RawMouseSessionRouter
        Private ReadOnly _engine As MousePerformanceEngine
        Private ReadOnly _analysisOptions As MousePerformanceAnalysisOptions
        Private ReadOnly _packetInbox As BufferedRawMousePacketInbox
        Private ReadOnly _packetPumpThread As Thread
        Private _acceptedSessionGeneration As Integer
        Private _captureActive As Integer
        Private _capturePumpsActive As Integer
        Private _packetEnqueuesInFlight As Integer
        Private _pendingQueuedPacket As QueuedRawMousePacket
        Private _packetConsumerDelayMilliseconds As Integer
        Private _selectedDeviceDisconnected As Boolean
        Private _disposed As Boolean

        Public Event DevicesChanged As EventHandler
        Public Event SelectedDeviceDisconnected As EventHandler

        Public Sub New(rawInputBroker As IRawInputBroker)
            Me.New(rawInputBroker, DefaultPacketBufferCapacity, 0, MousePerformanceAnalysisOptions.Default)
        End Sub

        Friend Sub New(rawInputBroker As IRawInputBroker,
                       packetBufferCapacity As Integer,
                       Optional packetConsumerDelayMilliseconds As Integer = 0,
                       Optional analysisOptions As MousePerformanceAnalysisOptions = Nothing)
            If rawInputBroker Is Nothing Then
                Throw New ArgumentNullException(NameOf(rawInputBroker))
            End If

            _rawInputBroker = rawInputBroker
            _sessionRouter = New RawMouseSessionRouter(_rawInputBroker, RawMouseSessionMode.SpecificDevice)
            _analysisOptions = If(analysisOptions, MousePerformanceAnalysisOptions.Default)
            _engine = New MousePerformanceEngine(_analysisOptions)
            _packetInbox = New BufferedRawMousePacketInbox(Math.Max(1, packetBufferCapacity))
            _packetConsumerDelayMilliseconds = Math.Max(0, packetConsumerDelayMilliseconds)
            _packetPumpThread = New Thread(AddressOf PacketPumpThreadMain) With {
                .IsBackground = True,
                .Name = "MousePerformancePacketPump"
            }

            AddHandler _rawInputBroker.MouseDevicesChanged, AddressOf OnDevicesChanged
            AddHandler _sessionRouter.PacketCaptured, AddressOf OnPacketCaptured
            AddHandler _sessionRouter.SelectedDeviceDisconnected, AddressOf OnSelectedDeviceDisconnected

            _packetPumpThread.Start()
        End Sub

        Public Function GetDevices() As IReadOnlyList(Of RawMouseDeviceInfo)
            Return FilterPhysicalDevices(_rawInputBroker.GetMouseDevices())
        End Function

        Public Function HasAvailableMouseDevice() As Boolean
            Dim devices = GetDevices()
            Return devices IsNot Nothing AndAlso devices.Count > 0
        End Function

        Public ReadOnly Property AnalysisOptions As MousePerformanceAnalysisOptions
            Get
                Return _analysisOptions
            End Get
        End Property

        Public ReadOnly Property CurrentSessionDeviceId As String
            Get
                SyncLock _syncRoot
                    Return _engine.SessionDeviceId
                End SyncLock
            End Get
        End Property

        Public Sub SetCpiState(effectiveCpi As Nullable(Of Double), canComputeVelocity As Boolean)
            SyncLock _syncRoot
                _engine.SetCpiState(effectiveCpi, canComputeVelocity)
            End SyncLock
        End Sub

        Public Function BeginSession(deviceId As String, startFresh As Boolean) As Boolean
            If String.IsNullOrWhiteSpace(deviceId) Then
                Return False
            End If

            Dim deviceExists = GetDevices().Any(Function(item) item IsNot Nothing AndAlso
                                                             String.Equals(item.DeviceId, deviceId, StringComparison.OrdinalIgnoreCase))
            If Not deviceExists Then
                Return False
            End If

            _selectedDeviceDisconnected = False
            Volatile.Write(_captureActive, 0)
            _packetInbox.Drain()

            _sessionRouter.BeginSession(deviceId)
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)

            SyncLock _syncRoot
                _engine.BeginCollecting(deviceId, Stopwatch.GetTimestamp(), startFresh)
            End SyncLock

            Volatile.Write(_capturePumpsActive, 1)
            Volatile.Write(_captureActive, 1)
            Return True
        End Function

        Public Sub PauseSession()
            _sessionRouter.PauseSession()

            ' Flush packets already captured by the previous session generation before we mark the engine as paused.
            ' This avoids losing the final button-only packet when right-click is used to exit collection.
            WaitForPacketEnqueueToSettle()
            FlushQueuedPackets()

            Volatile.Write(_captureActive, 0)
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)

            SyncLock _syncRoot
                _engine.PauseCollecting()
            End SyncLock
        End Sub

        Public Sub StopSession()
            _sessionRouter.StopSession()

            WaitForPacketEnqueueToSettle()
            FlushQueuedPackets()

            Volatile.Write(_captureActive, 0)
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)

            SyncLock _syncRoot
                _engine.StopCollecting()
            End SyncLock
        End Sub

        Public Sub ResetSession()
            Volatile.Write(_captureActive, 0)
            _sessionRouter.StopSession()
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)
            _selectedDeviceDisconnected = False

            SyncLock _syncRoot
                _engine.ResetSession()
            End SyncLock
        End Sub

        Public Function CaptureSnapshot(Optional includeEvents As Boolean = True) As MousePerformanceSnapshot
            Dim hasDevices = HasAvailableMouseDevice()
            SyncLock _syncRoot
                Return _engine.CreateSnapshot(ResolveStatus(hasDevices),
                                              Volatile.Read(_captureActive) <> 0,
                                              includeEvents)
            End SyncLock
        End Function

        Private Function ResolveStatus(hasDevices As Boolean) As MousePerformanceSessionStatus
            If _selectedDeviceDisconnected Then
                Return MousePerformanceSessionStatus.DeviceDisconnected
            End If

            If _engine.IsCollecting Then
                Return MousePerformanceSessionStatus.Collecting
            End If

            If _engine.IsFinalized Then
                Return MousePerformanceSessionStatus.Stopped
            End If

            If _engine.CanContinue Then
                Return MousePerformanceSessionStatus.Paused
            End If

            If Not hasDevices AndAlso Not _engine.HasData Then
                Return MousePerformanceSessionStatus.NoDevice
            End If

            Return MousePerformanceSessionStatus.Ready
        End Function

        Private Sub OnDevicesChanged(sender As Object, e As EventArgs)
            RaiseEvent DevicesChanged(Me, EventArgs.Empty)
        End Sub

        Private Sub OnPacketCaptured(sender As Object, e As RawMousePacketEventArgs)
            If e Is Nothing OrElse e.Packet Is Nothing Then
                Return
            End If

            If Volatile.Read(_captureActive) = 0 Then
                Return
            End If

            Dim acceptedGeneration = Volatile.Read(_acceptedSessionGeneration)
            If e.SessionGeneration <> acceptedGeneration Then
                Return
            End If

            Interlocked.Increment(_packetEnqueuesInFlight)
            Try
                Dim droppedCount = 0
                _packetInbox.Enqueue(e.Packet, acceptedGeneration, droppedCount)
                If droppedCount > 0 Then
                    SyncLock _syncRoot
                        _engine.ReportDroppedPackets(droppedCount)
                    End SyncLock
                End If
            Finally
                Interlocked.Decrement(_packetEnqueuesInFlight)
            End Try
        End Sub

        Private Sub OnSelectedDeviceDisconnected(sender As Object, e As EventArgs)
            WaitForPacketEnqueueToSettle()
            FlushQueuedPackets()

            Volatile.Write(_captureActive, 0)
            Volatile.Write(_acceptedSessionGeneration, _sessionRouter.SessionGeneration)
            _packetInbox.Drain()
            Volatile.Write(_capturePumpsActive, 0)
            _selectedDeviceDisconnected = True

            SyncLock _syncRoot
                _engine.PauseCollecting()
            End SyncLock

            RaiseEvent SelectedDeviceDisconnected(Me, EventArgs.Empty)
        End Sub

        Private Sub PacketPumpThreadMain()
            Do
                Dim queuedPacket As QueuedRawMousePacket = Nothing
                If Not _packetInbox.WaitDequeue(queuedPacket) Then
                    Exit Do
                End If

                Interlocked.Exchange(_pendingQueuedPacket, queuedPacket)

                SyncLock _queueProcessingSyncRoot
                    DrainQueuedPacketsUnsafe()
                End SyncLock
            Loop
        End Sub

        Private Sub FlushQueuedPackets()
            SyncLock _queueProcessingSyncRoot
                DrainQueuedPacketsUnsafe()
            End SyncLock
        End Sub

        Private Sub DrainQueuedPacketsUnsafe()
            Dim pendingQueuedPacket = Interlocked.Exchange(_pendingQueuedPacket, Nothing)
            If pendingQueuedPacket IsNot Nothing Then
                ProcessQueuedPacket(pendingQueuedPacket)
            End If

            Dim queuedPacket As QueuedRawMousePacket = Nothing
            While _packetInbox.TryDequeue(queuedPacket)
                ProcessQueuedPacket(queuedPacket)
            End While
        End Sub

        Private Sub ProcessQueuedPacket(queuedPacket As QueuedRawMousePacket)
            ProcessPacket(queuedPacket)
        End Sub

        Private Sub ProcessPacket(queuedPacket As QueuedRawMousePacket)
            If queuedPacket Is Nothing OrElse queuedPacket.Packet Is Nothing Then
                Return
            End If

            If _packetConsumerDelayMilliseconds > 0 Then
                Thread.Sleep(_packetConsumerDelayMilliseconds)
            End If

            If Volatile.Read(_captureActive) = 0 Then
                Return
            End If

            If queuedPacket.Generation <> Volatile.Read(_acceptedSessionGeneration) Then
                Return
            End If

            SyncLock _syncRoot
                _engine.PushPacket(queuedPacket.Packet)
            End SyncLock
        End Sub

        Private Sub WaitForPacketEnqueueToSettle(Optional timeoutMilliseconds As Integer = 1000)
            Dim timeout = Math.Max(0, timeoutMilliseconds)
            Dim deadline = Environment.TickCount64 + timeout

            While Volatile.Read(_packetEnqueuesInFlight) > 0
                If timeout > 0 AndAlso Environment.TickCount64 >= deadline Then
                    Exit While
                End If

                Thread.Sleep(1)
            End While
        End Sub

        Public Sub Dispose() Implements IDisposable.Dispose
            If _disposed Then
                Return
            End If

            _disposed = True
            Volatile.Write(_captureActive, 0)
            Volatile.Write(_capturePumpsActive, 0)

            _packetInbox.Dispose()
            If _packetPumpThread IsNot Nothing AndAlso _packetPumpThread.IsAlive Then
                _packetPumpThread.Join(1000)
            End If

            RemoveHandler _rawInputBroker.MouseDevicesChanged, AddressOf OnDevicesChanged
            RemoveHandler _sessionRouter.PacketCaptured, AddressOf OnPacketCaptured
            RemoveHandler _sessionRouter.SelectedDeviceDisconnected, AddressOf OnSelectedDeviceDisconnected
            _sessionRouter.Dispose()
        End Sub
    End Class
End Namespace

File: src/ClickSyncMouseTester/Services/MousePerformanceCaptureService.cs
#define TRACE
using ClickSyncMouseTester.Models;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Runtime.Versioning;
using System.Threading;

namespace ClickSyncMouseTester.Services;

[SupportedOSPlatform("windows")]
public class MousePerformanceCaptureService : IDisposable
{
    private sealed class QueueMetricsSnapshot
    {
        public int QueueOverflowCount { get; set; }

        public int QueueHighWatermark { get; set; }

        public int QueueCapacity { get; set; }
    }

    private const int DefaultPacketBufferCapacity = 65536;

    private const int PacketPumpShutdownJoinTimeoutMilliseconds = 1000;

    private readonly object _syncRoot;

    private readonly object _queueProcessingSyncRoot;

    private readonly object _queueMetricsSyncRoot;

    private readonly IRawMouseReportSource _mouseReportSource;

    private readonly IRawInputDeviceCatalog _deviceCatalog;

    private readonly RawMouseSessionRouter _sessionRouter;

    private readonly MousePerformanceEngine _engine;

    private readonly BufferedRawMousePacketInbox _packetInbox;

    private readonly Thread _packetPumpThread;

    private int _acceptedSessionGeneration;

    private int _captureActive;

    private int _packetEnqueuesInFlight;

    private int _packetConsumerDelayMilliseconds;

    private bool _selectedDeviceDisconnected;

    private int _hasAvailableMouseDevice;

    private int _queueOverflowCount;

    private MousePerformanceSnapshot _lastSummarySnapshot;

    private bool _disposed;

    public string CurrentSessionDeviceId
    {
        get
        {
            object syncRoot = _syncRoot;
            bool lockTaken = false;
            try
            {
                Monitor.Enter(syncRoot, ref lockTaken);
                return _engine.SessionDeviceId;
            }
            finally
            {
                if (lockTaken)
                {
                    Monitor.Exit(syncRoot);
                }
            }
        }
    }

    public event EventHandler DevicesChanged;

    public event EventHandler SelectedDeviceDisconnected;

    public MousePerformanceCaptureService(IRawInputBroker rawInputBroker)
        : this(rawInputBroker, DefaultPacketBufferCapacity, 0, MousePerformanceAnalysisOptions.Default)
    {
    }

    internal MousePerformanceCaptureService(IRawInputBroker rawInputBroker, int packetBufferCapacity, int packetConsumerDelayMilliseconds = 0, MousePerformanceAnalysisOptions analysisOptions = null)
        : this(rawInputBroker, rawInputBroker, packetBufferCapacity, packetConsumerDelayMilliseconds, analysisOptions)
    {
    }

    internal MousePerformanceCaptureService(IRawMouseReportSource mouseReportSource, IRawInputDeviceCatalog deviceCatalog, int packetBufferCapacity, int packetConsumerDelayMilliseconds = 0, MousePerformanceAnalysisOptions analysisOptions = null)
    {
        _syncRoot = new object();
        _queueProcessingSyncRoot = new object();
        _queueMetricsSyncRoot = new object();
        if (mouseReportSource == null)
        {
            throw new ArgumentNullException(nameof(mouseReportSource));
        }
        if (deviceCatalog == null)
        {
            throw new ArgumentNullException(nameof(deviceCatalog));
        }
        _mouseReportSource = mouseReportSource;
        _deviceCatalog = deviceCatalog;
        _sessionRouter = new RawMouseSessionRouter(_mouseReportSource, _deviceCatalog, RawMouseSessionMode.SpecificDevice);
        MousePerformanceAnalysisOptions analysisOptions2 = analysisOptions ?? MousePerformanceAnalysisOptions.Default;
        _engine = new MousePerformanceEngine(analysisOptions2);
        _packetInbox = new BufferedRawMousePacketInbox(Math.Max(1, packetBufferCapacity));
        _packetConsumerDelayMilliseconds = Math.Max(0, packetConsumerDelayMilliseconds);
        _packetPumpThread = new Thread(PacketPumpThreadMain)
        {
            IsBackground = true,
            Name = "MousePerformancePacketPump"
        };
        _deviceCatalog.MouseDevicesChanged += OnDevicesChanged;
        _sessionRouter.PacketCaptured += OnPacketCaptured;
        _sessionRouter.SelectedDeviceDisconnected += OnSelectedDeviceDisconnected;
        RefreshAvailableMouseDeviceState();
        _packetPumpThread.Start();
    }

    public IReadOnlyList<RawMouseDeviceInfo> GetDevices()
    {
        IReadOnlyList<RawMouseDeviceInfo> filteredDevices = MouseDeviceFiltering.FilterSelectableMotionDevices(_deviceCatalog.GetMouseDevices(), _deviceCatalog.GetMouseEndpointActivitySnapshots());
        UpdateAvailableMouseDeviceState(filteredDevices);
        return filteredDevices;
    }

    public bool HasAvailableMouseDevice()
    {
        return Volatile.Read(in _hasAvailableMouseDevice) != 0;
    }

    public void RequestDeviceRefresh(bool force = false)
    {
        _deviceCatalog.RequestMouseDevicesRefresh(force);
    }

    public void SetCpiState(double? effectiveCpi, bool canComputeVelocity)
    {
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.SetCpiState(effectiveCpi, canComputeVelocity);
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    public bool BeginSession(string deviceId, bool startFresh)
    {
        if (string.IsNullOrWhiteSpace(deviceId))
        {
            return false;
        }
        _deviceCatalog.RequestMouseDevicesRefresh(force: true);
        if (!MouseDeviceFiltering.ContainsSelectableMotionDevice(_deviceCatalog.GetMouseDevices(), _deviceCatalog.GetMouseEndpointActivitySnapshots(), deviceId))
        {
            return false;
        }
        _selectedDeviceDisconnected = false;
        Volatile.Write(ref _captureActive, 0);
        _packetInbox.Drain();
        Volatile.Write(ref _lastSummarySnapshot, null);
        if (startFresh)
        {
            ResetQueueMetrics();
        }
        _sessionRouter.BeginSession(deviceId);
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.BeginCollecting(deviceId, Stopwatch.GetTimestamp(), startFresh);
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
        Volatile.Write(ref _captureActive, 1);
        return true;
    }

    public void PauseSession()
    {
        _sessionRouter.PauseSession();
        WaitForPacketEnqueueToSettle();
        FlushQueuedPackets();
        Volatile.Write(ref _captureActive, 0);
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        _packetInbox.Drain();
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.PauseCollecting();
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    public void StopSession()
    {
        _sessionRouter.StopSession();
        WaitForPacketEnqueueToSettle();
        FlushQueuedPackets();
        Volatile.Write(ref _captureActive, 0);
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        _packetInbox.Drain();
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.StopCollecting();
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    public void ResetSession()
    {
        Volatile.Write(ref _captureActive, 0);
        _sessionRouter.StopSession();
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        WaitForPacketEnqueueToSettle();
        object queueProcessingSyncRoot = _queueProcessingSyncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(queueProcessingSyncRoot, ref lockTaken);
            _packetInbox.Drain();
            _selectedDeviceDisconnected = false;
            ResetQueueMetrics();
            object syncRoot = _syncRoot;
            bool lockTaken2 = false;
            try
            {
                Monitor.Enter(syncRoot, ref lockTaken2);
                _engine.ResetSession();
                Volatile.Write(ref _lastSummarySnapshot, null);
            }
            finally
            {
                if (lockTaken2)
                {
                    Monitor.Exit(syncRoot);
                }
            }
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(queueProcessingSyncRoot);
            }
        }
    }

    public MousePerformanceSnapshot CaptureSummarySnapshot()
    {
        return CaptureSnapshotCore(includeAnalysis: false);
    }

    public MousePerformanceSnapshot CaptureAnalysisSnapshot()
    {
        return CaptureSnapshotCore(includeAnalysis: true);
    }

    private MousePerformanceSnapshot CaptureSnapshotCore(bool includeAnalysis)
    {
        bool hasDevices = Volatile.Read(in _hasAvailableMouseDevice) != 0;
        QueueMetricsSnapshot queueMetricsSnapshot = GetQueueMetricsSnapshot();
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            if (!includeAnalysis && Volatile.Read(in _captureActive) != 0)
            {
                Monitor.TryEnter(syncRoot, ref lockTaken);
                if (!lockTaken)
                {
                    MousePerformanceSnapshot cachedSnapshot = Volatile.Read(ref _lastSummarySnapshot);
                    if (cachedSnapshot != null)
                    {
                        return cachedSnapshot;
                    }
                }
            }
            if (!lockTaken)
            {
                Monitor.Enter(syncRoot, ref lockTaken);
            }

            MousePerformanceSnapshot snapshot;
            if (includeAnalysis)
            {
                snapshot = _engine.CreateAnalysisSnapshot(ResolveStatus(hasDevices), Volatile.Read(in _captureActive) != 0, queueMetricsSnapshot.QueueOverflowCount, queueMetricsSnapshot.QueueHighWatermark, queueMetricsSnapshot.QueueCapacity);
            }
            else
            {
                snapshot = _engine.CreateSummarySnapshot(ResolveStatus(hasDevices), Volatile.Read(in _captureActive) != 0, queueMetricsSnapshot.QueueOverflowCount, queueMetricsSnapshot.QueueHighWatermark, queueMetricsSnapshot.QueueCapacity);
                Volatile.Write(ref _lastSummarySnapshot, snapshot);
            }
            return snapshot;
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    private MousePerformanceSessionStatus ResolveStatus(bool hasDevices)
    {
        if (_selectedDeviceDisconnected)
        {
            return MousePerformanceSessionStatus.DeviceDisconnected;
        }
        if (_engine.IsCollecting)
        {
            return MousePerformanceSessionStatus.Collecting;
        }
        if (_engine.IsFinalized)
        {
            return MousePerformanceSessionStatus.Stopped;
        }
        if (_engine.CanContinue)
        {
            return MousePerformanceSessionStatus.Paused;
        }
        if (!hasDevices && !_engine.HasData)
        {
            return MousePerformanceSessionStatus.NoDevice;
        }
        return MousePerformanceSessionStatus.Ready;
    }

    private void OnDevicesChanged(object sender, EventArgs e)
    {
        RefreshAvailableMouseDeviceState();
        DevicesChanged?.Invoke(this, EventArgs.Empty);
    }

    private void OnPacketCaptured(object sender, RawMousePacketEventArgs e)
    {
        if (e == null || e.Packet == null)
        {
            return;
        }
        Interlocked.Increment(ref _packetEnqueuesInFlight);
        try
        {
            if (Volatile.Read(in _captureActive) == 0)
            {
                return;
            }
            int acceptedSessionGeneration = Volatile.Read(ref _acceptedSessionGeneration);
            if (e.SessionGeneration != acceptedSessionGeneration)
            {
                return;
            }
            int droppedCount = 0;
            _packetInbox.Enqueue(e.Packet, acceptedSessionGeneration, ref droppedCount);
            if (droppedCount <= 0)
            {
                return;
            }
            RecordQueueOverflow(droppedCount);
            object syncRoot = _syncRoot;
            bool lockTaken = false;
            try
            {
                Monitor.Enter(syncRoot, ref lockTaken);
                _engine.ReportDroppedPackets(droppedCount);
            }
            finally
            {
                if (lockTaken)
                {
                    Monitor.Exit(syncRoot);
                }
            }
        }
        finally
        {
            Interlocked.Decrement(ref _packetEnqueuesInFlight);
        }
    }

    private void OnSelectedDeviceDisconnected(object sender, EventArgs e)
    {
        WaitForPacketEnqueueToSettle();
        FlushQueuedPackets();
        Volatile.Write(ref _captureActive, 0);
        Volatile.Write(ref _acceptedSessionGeneration, _sessionRouter.SessionGeneration);
        _packetInbox.Drain();
        _selectedDeviceDisconnected = true;
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.PauseCollecting();
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
        SelectedDeviceDisconnected?.Invoke(this, EventArgs.Empty);
    }

    private void PacketPumpThreadMain()
    {
        while (_packetInbox.WaitForData())
        {
            try
            {
                object queueProcessingSyncRoot = _queueProcessingSyncRoot;
                bool lockTaken = false;
                try
                {
                    Monitor.Enter(queueProcessingSyncRoot, ref lockTaken);
                    DrainQueuedPacketsUnsafe();
                }
                finally
                {
                    if (lockTaken)
                    {
                        Monitor.Exit(queueProcessingSyncRoot);
                    }
                }
            }
            catch (Exception ex)
            {
                Exception arg = ex;
                Trace.TraceError($"MousePerformance packet pump failed while draining the inbox: {arg}");
            }
        }
    }

    private void FlushQueuedPackets()
    {
        object queueProcessingSyncRoot = _queueProcessingSyncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(queueProcessingSyncRoot, ref lockTaken);
            DrainQueuedPacketsUnsafe();
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(queueProcessingSyncRoot);
            }
        }
    }

    private void DrainQueuedPacketsUnsafe()
    {
        QueuedRawMousePacket queuedPacket = QueuedRawMousePacket.Empty;
        while (_packetInbox.TryDequeue(ref queuedPacket))
        {
            ProcessPacket(queuedPacket);
        }
    }

    private void ProcessPacket(QueuedRawMousePacket queuedPacket)
    {
        if (!queuedPacket.HasPacket || Volatile.Read(in _captureActive) == 0 || queuedPacket.Generation != Volatile.Read(in _acceptedSessionGeneration))
        {
            return;
        }
        if (_packetConsumerDelayMilliseconds > 0)
        {
            Thread.Sleep(_packetConsumerDelayMilliseconds);
        }
        if (Volatile.Read(in _captureActive) == 0 || queuedPacket.Generation != Volatile.Read(in _acceptedSessionGeneration))
        {
            return;
        }
        object syncRoot = _syncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(syncRoot, ref lockTaken);
            _engine.PushPacket(queuedPacket.Packet);
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(syncRoot);
            }
        }
    }

    private void ResetQueueMetrics()
    {
        object queueMetricsSyncRoot = _queueMetricsSyncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(queueMetricsSyncRoot, ref lockTaken);
            _queueOverflowCount = 0;
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(queueMetricsSyncRoot);
            }
        }
        _packetInbox.ResetHighWatermark();
    }

    private void RecordQueueOverflow(int count)
    {
        if (count <= 0)
        {
            return;
        }
        object queueMetricsSyncRoot = _queueMetricsSyncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(queueMetricsSyncRoot, ref lockTaken);
            _queueOverflowCount += count;
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(queueMetricsSyncRoot);
            }
        }
    }

    private QueueMetricsSnapshot GetQueueMetricsSnapshot()
    {
        QueueMetricsSnapshot queueMetricsSnapshot = new QueueMetricsSnapshot();
        object queueMetricsSyncRoot = _queueMetricsSyncRoot;
        bool lockTaken = false;
        try
        {
            Monitor.Enter(queueMetricsSyncRoot, ref lockTaken);
            queueMetricsSnapshot.QueueOverflowCount = _queueOverflowCount;
        }
        finally
        {
            if (lockTaken)
            {
                Monitor.Exit(queueMetricsSyncRoot);
            }
        }
        queueMetricsSnapshot.QueueHighWatermark = _packetInbox.HighWatermark;
        queueMetricsSnapshot.QueueCapacity = _packetInbox.Capacity;
        return queueMetricsSnapshot;
    }

    private void RefreshAvailableMouseDeviceState()
    {
        UpdateAvailableMouseDeviceState(MouseDeviceFiltering.FilterSelectableMotionDevices(_deviceCatalog.GetMouseDevices(), _deviceCatalog.GetMouseEndpointActivitySnapshots()));
    }

    private void UpdateAvailableMouseDeviceState(IReadOnlyList<RawMouseDeviceInfo> devices)
    {
        Volatile.Write(ref _hasAvailableMouseDevice, (devices != null && devices.Count > 0) ? 1 : 0);
    }

    private void WaitForPacketEnqueueToSettle(int timeoutMilliseconds = 1000)
    {
        int timeout = Math.Max(0, timeoutMilliseconds);
        long deadlineTick = Environment.TickCount64 + timeout;
        while (Volatile.Read(ref _packetEnqueuesInFlight) > 0 && (timeout <= 0 || Environment.TickCount64 < deadlineTick))
        {
            Thread.Sleep(1);
        }
    }

    public void Dispose()
    {
        if (!_disposed)
        {
            _disposed = true;
            Volatile.Write(ref _captureActive, 0);
            _packetConsumerDelayMilliseconds = 0;
            _packetInbox.Complete();
            bool packetPumpStopped = true;
            if (_packetPumpThread != null && _packetPumpThread.IsAlive)
            {
                packetPumpStopped = _packetPumpThread.Join(PacketPumpShutdownJoinTimeoutMilliseconds);
            }
            if (packetPumpStopped)
            {
                _packetInbox.Dispose();
            }
            _deviceCatalog.MouseDevicesChanged -= OnDevicesChanged;
            _sessionRouter.PacketCaptured -= OnPacketCaptured;
            _sessionRouter.SelectedDeviceDisconnected -= OnSelectedDeviceDisconnected;
            _sessionRouter.Dispose();
        }
    }

    void IDisposable.Dispose()
    {
        this.Dispose();
    }
}





